site/guides/variant-constructor-pattern.md

```diff
@@ -0,0 +1,90 @@
+
+If you look at the steps involved in variant construction, it involves:
+1. set of functions that can help in updating unique identity of the resources
+2. Input to the functions for step 1
+
+[kpt-function-catalog](https://catalog.kpt.dev) provides two function that helps with customizing the identify of the resources:
+1. [set-namespace](https://catalog.kpt.dev/set-namespace/v0.3/): sets the namespace for all resources in the package.
+2. [ensure-name-substring](https://catalog.kpt.dev/ensure-name-substring/v0.2/): sets the name of the resources in the package.
+
+Now let's talk about the input to the functions. In most cases, variant's name (deployable package instance name) itself can be used to derive unique identity of the resources in the variant. For example, if I create a variant of `microservice` package, I will name the deployable package instance to `user-service` or `order-service`. So if the package's name is available to the functions, then they can use it to customize the name/namespace of the resources. So, starting with `kpt v1.0.0-beta.14+` , kpt makes `package's name` available in a local `ConfigMap` at a well-known path `pkg-context.yaml` that contains the package instance name and is available to functions during `kpt fn render` or `kpt fn eval`.
+
+Examples of `pkg-context.yaml` for abstract and deployable instance, For abstract package the value of `name` field is `example` and for deployable instance it contains actual value of the package name.
+
+```yaml
+# package context for an abstract package.
+# This is automatically created on `kpt pkg init`
+apiVersion: v1
+kind: ConfigMap
+data:
+	name: example
+```
+
+```yaml
+# package context for a deployable instance of a package. This is automatically
+# populated during `kpt pkg get`.
+apiVersion: v1
+kind: ConfigMap
+data:
+	name: my-pkg-instance
+```
+
+kpt supports a way to create `deployable pkg instance` such that `pkg-context` is automatically populated with the deployable instance's name.
+
+```sh
+$ kpt pkg get <pkg-location> my-pkg-instance --for-deployment
+```
+
+Now, let's look at how to provide the input to the functions.
+
+So if you are using `set-namespace` function in your package, then `set-namespace` function supports reading input from `pkg-context.yaml`. Here is an example:
+
+```Kptfile
+...
+pipeline:
+	mutators:
+		- img: set-namespace:v0.3.4
+		  configPath: pkg-context.yaml
+...
+```
+
+So for a package consumer, creating a deployable instance involves the following:
+
+```
+# pick name of the deployable instance say `my-pkg-instance`
+$ kpt pkg get <path-to-abstract-pkg> <my-pkg-instance> --for-deployement
+
+$ kpt fn render <my-pkg-instance>
+
+# That's it. Your pkg instance is all ready to deploy now.
```

Comment:
That may not be entirely it. It just customizes the names/namespace.

Other functions can use a similar pattern. For instance, we're using apply-replacements to set the RoleBinding group using the name of the package.

In some cases, the inputs needed to generate the variant will come from somewhere else. Those can be generated imperatively or manually edited after the package is forked. Additional customizations could also be made at that point.

Reply:
Yes, added a small paragraph discussing customization of non-identifier fields and rearranged the text.